Fontify ocamldoc comments

mattiase opened the pull request:
This makes markup constructs stand out in order to improve legibility
and reduce the risk of mistakes.

This may or may not be the best approach, but it does (mostly) work. Suggestions for improvement are welcome!

mattiase commented:
Oh, and a fix for `tuareg-font-lock-extension-node-face` got included half by mistake – that face looked terrible with a dark background (which I normally don't use but showed up in manual tests).

There should also be an ERT test for the change but I'd like to know if it's a good idea in general before wasting any more time.

mattiase commented:
Patch updated with various small improvements.

Many subtle mistakes in doc comments suddenly become very apparent. For example, stdlib/marshal.mli contains the construct `[[-2{^30}, 2{^30}-1]]` which is fontified very differently from the intended `\[-2{^30}, 2{^30}-1\]`.

monnier commented:
> This makes markup constructs stand out in order to improve legibility
> and reduce the risk of mistakes.
>
> This may or may not be the best approach, but it does (mostly)
> work. Suggestions for improvement are welcome!

Looks good, I think it's a good approach.
See some comments below,


        Stefan


> @@ -460,8 +460,12 @@ Valid names are `browse-url', `browse-url-firefox', etc."
>  
>  (defface tuareg-font-lock-extension-node-face
>    (if tuareg-faces-inherit-p
> -      '((t :inherit tuareg-font-lock-infix-extension-node-face
> -           :background "gray92"))
> +      '((((background light))
> +         (:inherit tuareg-font-lock-infix-extension-node-face
> +                   :background "gray92"))
> +        (((background dark))
> +         (:inherit tuareg-font-lock-infix-extension-node-face
> +                   :foreground "LightSteelBlue")))
>      '((((background light)) (:foreground "Orchid" :background "gray92"))
>        (((background dark)) (:foreground "LightSteelBlue" :background "gray92"))
>        (t (:foreground "LightSteelBlue"))))

LGTM (no opinion about the color), but you can use something like

    '((default :inherit tuareg-font-lock-infix-extension-node-face)
      (((background dark)) :background "LightSteelBlue")
      (t :background "gray92")

which avoids the :inherit duplication and make sure we still have the
old color for non-dark non-light backgrounds (not sure if those can
occur, but with this version I don't need to worry about it).

> @@ -586,7 +600,7 @@ Regexp match data 0 points to the chars."
>      (if (or (eq (char-syntax (or (char-before mbegin) ?\ )) syntax)
>              (eq (char-syntax (or (char-after mend) ?\ )) syntax)
>              (memq (get-text-property mbegin 'face)
> -                  '(tuareg-doc-face
> +                  '(font-lock-doc-face
>                      font-lock-string-face
>                      font-lock-comment-face
>                      tuareg-font-lock-error-face

What happened to poor `tuareg-doc-face`?

> +(defun tuareg-fontify-doc-comment (state)
> +  (let ((beg (nth 8 state))
> +        (end (save-excursion
> +               (parse-partial-sexp (point) (point-max) nil nil state
> +                                   'syntax-table)
> +               (point))))
> +    (put-text-property beg end 'face 'font-lock-doc-face)
> +    (save-excursion
> +      (let ((case-fold-search nil)
> +            (bracket-depth 0)           ; [] nesting level
> +            (verbatim nil))             ; inside {v ... v}
> +        (goto-char beg)
> +        (while (< (point) end)
> +          (let ((start (point)))
> +            (cond
> +             ;; Inside [].
> +             ((> bracket-depth 0)
> +              (if (re-search-forward (rx (or (in "[]") (seq "\\" anything)))
> +                                     end t)
> +                  (let ((next (char-after (match-beginning 0))))
> +                    (put-text-property start (point) 'face
> +                                       tuareg-font-lock-doc-code-face)
> +                    (cond
> +                     ((eq next ?\[)
> +                      (setq bracket-depth (1+ bracket-depth)))
> +                     ((eq next ?\])
> +                      (setq bracket-depth (1- bracket-depth))
> +                      (when (= bracket-depth 0)
> +                        (put-text-property
> +                         (1- (point)) (point) 'face
> +                         'tuareg-font-lock-doc-markup-face)))))
> +                (put-text-property start end 'face
> +                                   tuareg-font-lock-doc-code-face)
> +                (goto-char end)))

IIUC you don't really try to handle nesting (as in ocamldoc comments
within code within an ocamldoc comment), so I think you can use a loop
whose body always starts "outside of any ocamldoc markup" and where you
just search for "\\[\\|{v\\|...\\|..." and then process that thing before
returning to the beginning of the loop.

To find the matching ']' you can use something like

    (with-syntax-table tuareg--fontify-ocamldoc-code-syntax-table
      (up-list 1))

so you don't have to count them by hand.

> +             ((eq (following-char) ?\[)
> +              (put-text-property start (1+ start) 'face
> +                                 'tuareg-font-lock-doc-markup-face)
> +              (setq bracket-depth 1)
> +              (forward-char 1))

I.e. here instead of looping back, do the `up-list` dance right away.

You might be able to reuse part of `syntax-propertize-rules` to generate
the "\\[\\|{v\\|...\\|..." and the dispatch to the corresponding cases
from source code that looks a bit like "lex".


        Stefan

mattiase commented:
> '((default :inherit tuareg-font-lock-infix-extension-node-face)
> (((background dark)) :background "LightSteelBlue")
> (t :background "gray92")
> 
> which avoids the :inherit duplication and make sure we still have the
> old color for non-dark non-light backgrounds (not sure if those can
> occur, but with this version I don't need to worry about it).

Thank you, that is indeed what I wanted to do; I've changed to using that expression.
I'm quite sure the new colours are more or less what was intended all along; lightsteelblue-on-gray92 makes no sense for those unfortunate to have the poor photoreceptors of Earthlings.

(Seems the Github mailer smashes indentation by the way. Maybe using Markdown triple-grave quoting would help?)

> > (memq (get-text-property mbegin 'face)
> > - '(tuareg-doc-face
> > + '(font-lock-doc-face
> > font-lock-string-face
> > font-lock-comment-face
> > tuareg-font-lock-error-face
> 
> What happened to poor `tuareg-doc-face`?

It's just a variable, not a face, so the previous code could never have worked in the first place. The value was always `font-lock-doc-face`. I've remove `tuareg-doc-face` entirely since it didn't really carry any weight. We could turn it into a proper `defface` if that would help somehow.

In fact, the extra `defvar` for each `defface` in the file seems very much like long obsolete practice -- at least that is what the comment at font-lock.el:292 appears to say. I didn't carry on the tradition in the two new faces added by the patch.

And isn't `tuareg-faces-inherit-p` always true? `face-attribute-name-alist` had `:inherit` long before Emacs 24.4, which is the minimum version for tuareg. Simplifications to be made!

(There is no standard face for doc markup, and I'm apparently not the first to have noticed: cc-fonts.el will use `font-lock-doc-markup-face` if available. Maybe it's something to add to Emacs?)

> IIUC you don't really try to handle nesting (as in ocamldoc comments
> within code within an ocamldoc comment), so I think you can use a loop
> whose body always starts "outside of any ocamldoc markup" and where you
> just search for "\\[\\|{v\\|...\\|..." and then process that thing before
> returning to the beginning of the loop.

Thank you and yes, you are perfectly right; I've now done some restructuring. (I considered carrying your suggestion to the limit and put everything into a single giant regex with submatches for each case but it just became too unwieldy.)

> To find the matching ']' you can use something like
> 
> (with-syntax-table tuareg--fontify-ocamldoc-code-syntax-table
> (up-list 1))
> 
> so you don't have to count them by hand.

That's not a bad idea! However given the need to set up a new syntax table (admittedly a simple one), the need for narrowing to limit the scope of up-list, catching exceptions in case the brackets don't balance etc, I'd say it's a wash in terms of complexity and likely slower overall given all that up-list has to contend with. I could be wrong!

> You might be able to reuse part of `syntax-propertize-rules` to generate
> the "\\[\\|{v\\|...\\|..." and the dispatch to the corresponding cases
> from source code that looks a bit like "lex".

Hmmm... do you mean that we could somehow run the standard fontification on the inside of [...] instead of just giving it the nil face? Right now, those parts are hit-and-miss, mainly (I think) depending on the context of the code preceding the doc comment.

I also fixed a number of unexpected little details in the ocamldoc syntax: backslashes only escape `{}[]@` and not, say, `<` or other backslashes, which means that `[hello\\]]` is properly balanced.

monnier commented:
> (Seems the Github mailer smashes indentation by the way. Maybe using
> Markdown triple-grave quoting would help?)

For some reason, Github doesn't want to accept markdown via email (I
tried to use a `Content-Type: text/x-markdown` but it makes it worse:
Github then rejects the message altogether).

>> What happened to poor `tuareg-doc-face`?
> It's just a variable, not a face, so the previous code could never have
> worked in the first place.

Ah, that makes sense.
But the patch doesn't remove that var, right?  Should it mark it as obsolete?

> And isn't `tuareg-faces-inherit-p` always true? `face-attribute-name-alist`
> had `:inherit` long before Emacs 24.4, which is the minimum version for
> tuareg. Simplifications to be made!

:inherit dates back to Emacs-21 but I'm not sure if/when XEmacs got it.

> (There is no standard face for doc markup, and I'm apparently not the first
> to have noticed: cc-fonts.el will use `font-lock-doc-markup-face` if
> available. Maybe it's something to add to Emacs?)

Might be a good idea, yes.

> Thank you and yes, you are perfectly right; I've now done some
> restructuring. (I considered carrying your suggestion to the limit and put
> everything into a single giant regex with submatches for each case but it
> just became too unwieldy.)

That's why I suggested going the `syntax-propertize-rules` way.

> That's not a bad idea! However given the need to set up a new syntax table
> (admittedly a simple one), the need for narrowing to limit the scope of
> up-list, catching exceptions in case the brackets don't balance etc, I'd say
> it's a wash in terms of complexity and likely slower overall given all that
> up-list has to contend with. I could be wrong!

I haven't measured it, so you might be right.

>> You might be able to reuse part of `syntax-propertize-rules` to generate
>> the "\\[\\|{v\\|...\\|..." and the dispatch to the corresponding cases
>> from source code that looks a bit like "lex".
> Hmmm... do you mean that we could somehow run the standard fontification on
> the inside of [...] instead of just giving it the nil face?

No, I'm suggesting you look at the definition of
`syntax-propertize-rules` and plagiarize the part that takes a bunch of
regexps and associated code and generates the giant regexps (and the
code to dispatch based on submatch).


        Stefan

mattiase commented:
> >> What happened to poor `tuareg-doc-face`?
> > It's just a variable, not a face, so the previous code could never have
> > worked in the first place.
> 
> Ah, that makes sense.
> But the patch doesn't remove that var, right? Should it mark it as obsolete?

The updated patch did remove it. I don't know if it should be marked as obsolete; it doesn't seem to be a very obvious variable to rely on from user code.

> > And isn't `tuareg-faces-inherit-p` always true? `face-attribute-name-alist`
> > had `:inherit` long before Emacs 24.4, which is the minimum version for
> > tuareg. Simplifications to be made!
> 
> :inherit dates back to Emacs-21 but I'm not sure if/when XEmacs got it.

I thought tuareg no longer worked with XEmacs, or does it? Looks like you removed support for it last year in d7c8618955bd461c5ffadbceecd6f4b5565a5f12.

Besides, keeping it working with XEmacs just isn't practical; I don't test on it, the CI didn't either (back when it was still running...) and it's hard to know exactly what constructs we can or cannot use.

> > (There is no standard face for doc markup, and I'm apparently not the first
> > to have noticed: cc-fonts.el will use `font-lock-doc-markup-face` if
> > available. Maybe it's something to add to Emacs?)
> 
> Might be a good idea, yes.

Preferably it should be made to work with existing themes and other user customisations without beastly results.

> No, I'm suggesting you look at the definition of
> `syntax-propertize-rules` and plagiarize the part that takes a bunch of
> regexps and associated code and generates the giant regexps (and the
> code to dispatch based on submatch).

Okay, I'm updating the PR with several changes to give you an idea what the different approaches would look like.
First the old sequence of multiple regexp matching constructs. Then, a hand-written and hand-optimised single-big-regexp version. Finally, a big-regexp version generated by a macro similar to `syntax-propertize-rules` but much simpler.

The last version is pretty but has the highest line count.
I haven't seen any obvious difference in performance.

monnier commented:
>> Ah, that makes sense.
>> But the patch doesn't remove that var, right? Should it mark it as obsolete?
> The updated patch did remove it.

Perfect.

> I don't know if it should be marked as obsolete; it doesn't seem to be
> a very obvious variable to rely on from user code.

No, probably not worth it.  Of course, if someone complains, we can
revisit this choice.

>> > And isn't `tuareg-faces-inherit-p` always true? `face-attribute-name-alist`
>> > had `:inherit` long before Emacs 24.4, which is the minimum version for
>> > tuareg. Simplifications to be made!
>> :inherit dates back to Emacs-21 but I'm not sure if/when XEmacs got it.
> I thought tuareg no longer worked with XEmacs, or does it?

Not that I know.

>> > (There is no standard face for doc markup, and I'm apparently not the first
>> > to have noticed: cc-fonts.el will use `font-lock-doc-markup-face` if
>> > available. Maybe it's something to add to Emacs?)
>> Might be a good idea, yes.
> Preferably it should be made to work with existing themes and other user
> customisations without beastly results.

It just needs a reasonable default value for the different cases
(low/bi-color screens, dark/light backgrounds), but that shouldn't be
too hard (we can even start with something naive and let others improve
it).

> Okay, I'm updating the PR with several changes to give you an idea what the
> different approaches would look like.

I just pushed the branch.  I think `tuareg--syntax-rules` can be
improved a bit further and then added to Emacs itself (so we can get rid
of it in Tuareg in a few years).

Things I'd like to see there are:
- make it work for non-rx regexps as well (in `syntax-propertize-rules`
  I did that by `eval`ing the regexps).
- get rid of `start` and `group` variables since it's not hygienic.
  `start` can be replaced by (match-beginning 0) anyway, and `group` can
  be eliminated by shifting the match data accordingly.
- ideally, `syntax-propertize-rules` can be changed to make use of the
  new macro.


        Stefan

mattiase commented:
> It just needs a reasonable default value for the different cases
> (low/bi-color screens, dark/light backgrounds), but that shouldn't be
> too hard (we can even start with something naive and let others improve
> it).

I suppose we could define it as :inherit font-lock-constant-face by default, since that is what cc-mode does (and I copied for Tuareg)?

It might be useful to consult authors of other programming modes and/or themes about what would be a good style and for what kind of syntactic construct.

> Things I'd like to see there are:
> - make it work for non-rx regexps as well (in `syntax-propertize-rules`
> I did that by `eval`ing the regexps).

Actually it works for non-rx regexps; I just failed to update the commit message (sorry). It's done by macroexpanding the regexps. (I started writing something much fancier that relied on regexp transformations which was much easier in rx form, but ended up doing it the simple way.)

> - get rid of `start` and `group` variables since it's not hygienic.
> `start` can be replaced by (match-beginning 0) anyway, and `group` can
> be eliminated by shifting the match data accordingly.

Yes, `start` was just a minor convenience. Shifting the match data was one thing I didn't like about `syntax-propertize-rules`; I'd rather use lambdas for the clause bodies and pass `group` as a parameter.

> - ideally, `syntax-propertize-rules` can be changed to make use of the
> new macro.

Yes, that might be feasible.

Another open question: the current fontification accepts any ocamldoc @-tag name, since the docs say that users can add custom tags. However I've found little evidence for that feature being used in practice so perhaps it's more helpful to check against a list of predefined tags so that we immediately detect mistakes such as writing ***@***.***` instead of ***@***.***`, something that isn't treated as a hard error by ocamldoc but merely results in the clause being omitted from the output (HTML etc).

mattiase commented:
> mistakes such as writing `***@***.***` instead of `***@***.***`

Looks like github's mailer is at it again. This should be

[...] mistakes such as writing `@returns` instead of `@return` [...]

monnier commented:
> Yes, `start` was just a minor convenience. Shifting the match data was one
> thing I didn't like about `syntax-propertize-rules`;

But that's an implementation detail.  I think we should focus on the
cleanliness of the API.  We can later change the implementation into one
that doesn't need to shift anything (once we have a better regexp
engine ;-).

> Another open question: the current fontification accepts any ocamldoc @-tag
> name, since the docs say that users can add custom tags. However I've found
> little evidence for that feature being used in practice so perhaps it's more
> helpful to check against a list of predefined tags so that we immediately
> detect mistakes such as writing ***@***.***` instead of ***@***.***`,
> something that isn't treated as a hard error by ocamldoc but merely results
> in the clause being omitted from the output (HTML etc).

I have no strong opinion about that, so as the coder you'll get to
choose.

mattiase commented:
> I have no strong opinion about that, so as the coder you'll get to choose.

OK, that's now pull request #258.
